crates/component/src/wrappers/http.rs

```diff
@@ -0,0 +1,570 @@
+        //};
+        //let trailers = trailers.map_err(|()| std::io::Error::from(std::io::ErrorKind::Other))?;
+        //let trailers =
+        //    trailers.map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
+        wasi::http::types::OutgoingBody::finish(body, None)
+            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))
+    }
+}
+
+impl OutgoingBody for IncomingBody {
+    fn write(
+        self,
+        body: wasi::http::types::OutgoingBody,
+        stream: OutputStream,
+    ) -> std::io::Result<()> {
+        loop {
+            match stream.blocking_splice(&self.stream, u64::MAX) {
+                Ok(0) | Err(StreamError::Closed) => break,
+                Ok(_) => continue,
+                Err(StreamError::LastOperationFailed(err)) => {
+                    return Err(std::io::Error::new(std::io::ErrorKind::Other, err));
+                }
+            }
+        }
+        let trailers = self
+            .into_trailers_wasi()
+            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
+        wasi::http::types::OutgoingBody::finish(body, trailers)
+            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))
+    }
+}
+
+/// Wraps a body, which implements [Read]
+#[derive(Clone, Copy, Debug)]
+pub struct ReadBody<T>(T);
+
+impl<T: Read> OutgoingBody for ReadBody<T> {
+    fn write(
+        mut self,
+        body: wasi::http::types::OutgoingBody,
+        mut stream: OutputStream,
+    ) -> std::io::Result<()> {
+        std::io::copy(&mut self.0, &mut stream)?;
+        drop(stream);
+        wasi::http::types::OutgoingBody::finish(body, None)
+            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))
+    }
+}
+
+pub struct IncomingBody {
+    body: wasi::http::types::IncomingBody,
+    stream: InputStream,
+}
+
+impl TryFrom<wasi::http::types::IncomingBody> for IncomingBody {
+    type Error = anyhow::Error;
+
+    fn try_from(body: wasi::http::types::IncomingBody) -> Result<Self, Self::Error> {
+        let stream = body
+            .stream()
+            .map_err(|()| anyhow!("failed to get incoming request body"))?;
+        Ok(Self { body, stream })
+    }
+}
+
+impl TryFrom<wasi::http::types::IncomingRequest> for IncomingBody {
+    type Error = anyhow::Error;
+
+    fn try_from(request: wasi::http::types::IncomingRequest) -> Result<Self, Self::Error> {
+        let body = request
+            .consume()
+            .map_err(|()| anyhow!("failed to consume incoming request"))?;
+        body.try_into()
+    }
+}
+
+impl Deref for IncomingBody {
+    type Target = InputStream;
+
+    fn deref(&self) -> &Self::Target {
+        &self.stream
+    }
+}
+
+impl DerefMut for IncomingBody {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.stream
+    }
+}
+
+impl From<IncomingBody> for InputStream {
+    fn from(IncomingBody { stream, .. }: IncomingBody) -> Self {
+        stream
+    }
+}
```

Comment:
Sadly I don't think we can implement this, because dropping the body before the input stream will cause a wasmtime panic